### ISSUE TITLE
Do not try and eta reduce runST

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/hlint#136]
From http://code.google.com/p/ndmitchell/issues/detail?id=618:

hlint-1.8.49 when applied to this code:

```
import Control.Monad.ST

action :: Int -> ST s Int
action x = return (x + 1)

tmp :: [Int] -> [Int]
tmp = map (\ x -> runST $ action x)
```

gives:

```
Warning: Avoid lambda
Found:
  \ x -> runST $ action x
Why not:
  runST . action
```

runST has a polymorphic type of rank 2, so such modification is incorrect due to problem with (.) typization and will cause compilation error.

I replied: runST is _the_ rank2 function, so maybe I should just not do anything when I see a runST and be super careful.

**Comment by mishun** [comment]
The actual problem here was with custom rank-2 functions: runST is not the only example. As I could understand it is very hard to fix properly since hlint does not perform complete typecheck. Maybe add special comments understandable by hlint like {- hlint-notify-higher-rank: runST -}?

**Comment by ndmitchell** [comment]
For top level things, if the variable is named `mr` then I don't eta reduce it. I wonder if I should add the same rule for lambdas? Would `\mr -> ...` be excessively hideous? Marking some functions as sensitive isn't out of the question either.

**Comment by mishun** [comment]
I got a bit confused: transformation in example above is actually not a η-reduction (although one of its steps is η-reduction) and it is not η-reduction that causes the problem.
```Haskell
\ x -> ($) runST (action x)
```
is β-equivalent to:
```Haskell
\ x -> runST (action x)
```
then we substitute definition (.) f g x = f (g x) and get:
```Haskell
\ x -> (.) runST action x
```
and finally η-reduce to:
```Haskell
(.) runST action
```
η-reduction here is perfectly fine --- it is 2-nd step that causes the problem (because (.) receives rank-3 type).

Looks like hlint treats this transformation separately from η-reduction and exclusively for λ-expressions, as:
```Haskell
f x = g $ h x
f x = g (h x)
```
get no suggestion, but
```Haskell
\ x -> g $ h x
\ x -> g (h x)
```
get something like:
```
Suggestion: Avoid lambda
Found:
  \ x -> g (h x)
Why not:
  g . h
```

**Comment by mishun** [comment]
Simple η-reduction is treated differently for lambas too:
```
Warning: Eta reduce
Found:
  f x = h x
Why not:
  f = h

Warning: Avoid lambda
Found:
  \ x -> h x
Why not:
  h
```

**Comment by ndmitchell** [comment]
The idea is that if you can eliminate a lambda you are making the code noticeably simpler, while just shaving an argument off a definition isn't as useful a reduction, so it has to be slightly simpler. It's definitely not a strict divide.

**Comment by ndmitchell** [comment]
I have special cased runST, and am happy to special case other such functions if they occur.
